scylla-cdc/src/reader.rs

```diff
@@ -0,0 +1,266 @@
+use std::cmp::{max, min};
+use std::error::Error;
+use std::time;
+use std::time::SystemTime;
+
+use chrono;
+use futures::stream::StreamExt;
+use scylla::frame::value::Timestamp;
+use scylla::{Session, SessionBuilder};
+use tokio::time::sleep;
+
+type StreamID = Vec<u8>;
+
+pub struct Reader {
+    stream_id_vec: Vec<StreamID>,
+    lower_timestamp: chrono::Duration,
+    window_size: chrono::Duration,
+    safety_interval: chrono::Duration,
+    upper_timestamp: Option<chrono::Duration>,
+    sleep_interval: time::Duration,
+}
+
+impl Reader {
+    pub fn new(
+        stream_ids: Vec<StreamID>,
+        start_timestamp: chrono::Duration,
+        window_size: chrono::Duration,
+        safety_interval: chrono::Duration,
+        sleep_interval: time::Duration,
+    ) -> Reader {
+        Reader {
+            stream_id_vec: stream_ids,
+            lower_timestamp: start_timestamp,
+            window_size,
+            safety_interval,
+            upper_timestamp: std::option::Option::None,
+            sleep_interval,
+        }
+    }
+
+    // Mutable access.
+    fn lower_timestamp_mut(&mut self) -> Option<chrono::Duration> {
+        self.upper_timestamp
+    }
+
+    pub fn set_lower_timestamp(&mut self, new_lower_timestamp: chrono::Duration) {
+        self.upper_timestamp = Some(new_lower_timestamp);
+    }
```

Comment:
While writing the issue's description I might have been imprecise in phrasing the point about setting the end timestamp for the reader - I meant that both options should be possible: setting the timestamp before the reader is running AND while it is running. However, the current API does not allow for that because of Rust's ownership rules.

- `Reader::fetch_cdc` which runs the reading process borrows `Reader` by immutable reference (`&self`),
- `Reader::set_lower_timestamp` borrows `Reader` by mutable reference (`&mut self`).

Because Rust does not allow a mutable reference and immutable reference to the same object exist at the same time, it makes it impossible to set the timestamp while `fetch_cdc` is running.

There is a way out of this issue - you can use a Mutex to protect the lower_timestamp:

```rust
use tokio::sync::Mutex;

pub struct Reader {
    // ...
    upper_timestamp: tokio::sync::Mutex<Option<chrono::Duration>>,
    // ...
}
```

...and `set_lower_timestamp` then may be changed to take an immutable reference - Rust allows multiple immutable references to the same objects to exist:

```rust
pub fn set_lower_timestamp(&mut self, new_lower_timestamp: chrono::Duration) {
    let guard = self.upper_timestamp.lock().await;
    *guard = Some(new_lower_timestamp);
}
```

Adjusting `fetch_cdc` is an exercise for the reader :)

Reply:
Oh, I just noticed that this function is named `set_lower_timestamp`, but it sets the _upper_ timestamp. The function's name should be changed to `set_upper_timestamp`.

Reply:
^Yeah, this problem was already spotted ;)